Set DPI scaling on first launch

tobiaskohlbau opened the pull request:
When OpenRCT2 is first launched every setting gets defaulted.
DPI default is set to 1.0 which is to small for high dpi monitors.
It's not a big issue as a scaling is already supported but hitting
the tremendous small upwards icon is hard. After this patch
on first launch the monitor scaling is checked and set to a more
appropriate value.

ccoors commented:
Some time ago I wrote a patch for Linux implementing that, if you want you can include it. It's not perfect but it's a start.

[dpi.zip](https://github.com/OpenRCT2/OpenRCT2/files/1620320/dpi.zip)

rwjuk commented:
Here's how to do it for macOS - I can implement this at some point if you're not used to macOS https://stackoverflow.com/questions/12589198/how-to-read-the-physical-screen-size-of-osx/12589799#12589799

janisozaur commented:
Unfortunately MinGW does not ship `ShellScalingAPI.h` file nor `GetDpiForSystem` function in any other header. You'll have to remove it from compilation for MinGW.

janisozaur commented:
@ccoors I recall we discussed the SDL2-provided functions and they proved to be insufficient for some reason, do you remember what it was?

ccoors commented:
@janisozaur Yes, the settings are read long before SDL is initialized :/ Otherwise we could use SDL on all platforms.

rwjuk commented:
Could we just reset the scaling once SDL is initialised?

ccoors commented:
Set some kind of flag so it's updated once SDL is initilized if there was no previous setting? Seems to complicate things quite a bit but might actually be better than implementing platform-specific code.

janisozaur commented on src/openrct2/platform/windows.c:
```diff
@@ -1005,4 +1007,15 @@ bool platform_setup_uri_protocol()
+    uint32 dpi = GetDpiForSystem();
+    if (dpi > 96)
+    {
+        return 2.0;
```
Why can't you just return `dpi / 96.f`? We have "Smooth NN" (#6588) implemented which provides nice user experience even on non-integer scales and we ship it this way on Android.

I also think returning just two values is rather limiting, devices like https://www.notebookcheck.net/HP-Spectre-x360-13-7500U-4K-UHD-Convertible-Review.210812.0.html have 13" 2160p screens, that's around 330ppi, so easily that should at least be 3 to be usable in such case.

ccoors commented:
I'm against setting fractures as the default Software renderer can't do smoothing. And even then, it destroys the retro look (imho).

tobiaskohlbau commented:
@ccoors I'm against fractures also cause if the fraction is for e.g. `1.25` the user increments in settings by `0.5`. They're never able to set it back to clean state without editing config by hand.

I've implemented a platform independent check for DPI scaling on first launch now. In addition DPI is scaled upon monitor dpi but in steps of `0.5` to align with current possible values.

Please add the label `testing required`. (I will spin up my old Android tablet so see if it works there)

janisozaur commented:
> if the fraction is for e.g. 1.25 the user increments in settings by 0.5. They're never able to set it back to clean state without editing config by hand.

That's not entirely correct. User increments are 0.25 and you can modify it to arbitrary value with `set window_scale X`

tobiaskohlbau commented:
Ok bad example what if `1.13331` would be the calculated result. Which non techy uses cli commands to set window scale :-P

marijnvdwerf commented:
Current Android behavior is as-intended. Scaling should not be rounded to 0.25f increments.

@rwjuk Current builds already work fine with Retina support.

marijnvdwerf commented on src/openrct2-ui/UiContext.cpp:
```diff
@@ -96,6 +97,23 @@ class UiContext final : public IUiContext
+            {
+                SDLException::Throw("SDL_GetDisplayDPI(...)");
+            }
+            gConfigGeneral.window_scale = std::round(4 * ddpi) / 384.0f;
```
Magic number.

tobiaskohlbau commented:
I've rebased this PR and changed scaling to not round for any device as by public demand.

@marijnvdwerf different pixel densities per screen are not supported at the moment and neither are after this PR. It's only a fix for first launch at high dpi where it is annoying to hit the tremendous small arrows.

If they currently work within Retina I would assume there is some kind of scaling by OS as supported by Windows as well but it's not best behavior if there exist specific scaling options.

AaronVanGeffen commented:
I've rebased this PR on top of current develop.

@marijnvdwerf Is the PR in its current state acceptable for merging, from your point of view?

AaronVanGeffen commented on src/openrct2-ui/UiContext.cpp:
```diff
@@ -103,6 +104,24 @@ class UiContext final : public IUiContext
+        {
+#ifndef __MINGW32__
+            float ddpi, hdpi, vdpi;
+            if (SDL_GetDisplayDPI(0, &ddpi, &hdpi, &vdpi))
```
This should store its (return) error code, and pass it to [SDL_GetError](https://wiki.libsdl.org/SDL_GetError). That way it'll provide us with some useful error information, rather than just crashing with an unspecified uncaught exception.

AaronVanGeffen commented:
When I run this on Linux, it sets `window_scale = 0.995974` for my 96dpi screen…

I'm going to close this, seeing as it appears to have been abandoned, does not behave properly, and has sparked more controversy than agreement.
